### PR TITLE
Arizona - exclusion for state or local government pensions

### LINKED ISSUE #3048: Arizona- Exclusion for State or Local Government Pensions
reference:
https://azdor.gov/sites/default/files/2023-03/FORMS_INDIVIDUAL_2022_140BOOKLET.pdf (Line 29a)
https://www.azleg.gov/ars/43/01022.htm

### PR DESCRIPTION
**Pull request by phywzc**
Fixes #3048

### Summary
📄🧮🏛️

This pull request adds a new feature to the policyengine_us model that allows simulating the effect of Arizona's pension exclusion policy on state income tax. It introduces a new parameter file `maximum_amount.yaml` and a new variable file `az_pension_exclusion.py` to implement the policy logic.

> _New parameter_
> _`pension_exclusion` added_
> _Autumn of one's life_

### Walkthrough
*  Add a new parameter and a new variable for Arizona pension exclusion ([link](https://github.com/PolicyEngine/policyengine-us/pull/3050/files?diff=unified&w=0#diff-5e05ecf21e391e7b46dccae5c735bb4343384c6608dded670b25d491c3870d70R1-R12), [link](https://github.com/PolicyEngine/policyengine-us/pull/3050/files?diff=unified&w=0#diff-9ec394a990080b738685dadb4c0377f2485549ecc0ad0ff3abfc0c63c0729387L1-R32))

### CODE REVIEW

File: policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+description: Arizona allows an exclusion of taxable pension income to get its net income that is capped at this amount.
+metadata:
+  label: Arizona maximum pension exclusion amount
+  period: year
+  unit: currency-USD
+  reference:
+  - title: 2022 Arizona Form 140 Resident Personal Income Tax Booklet
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml:7]: add legal code

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml:7]: https://www.azleg.gov/ars/43/01022.htm

File: policyengine_us/variables/gov/states/az/tax/income/net_income/az_pension_exclusion.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    defined_for = StateCode.AZ 
+
+    def formula(tax_unit, period, parameters):
+        p = parameters(period).gov.states.az.tax.income.pension_exclusion
+        person = tax_unit.members
+        filing_status = tax_unit("filing_status", period)
+        is_joint = filing_status == filing_status.possible_values.JOINT
+
+        # calculate the total allowed pension
+        total_allowed_pension_exclusion = min_(p.maximum_amount, person("public_pension_income", period))
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/net_income/az_pension_exclusion.py]: taxable_public_pension_income

Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/net_income/az_pension_exclusion.py]:
```suggestion
        pension_income = person("public_pension_income", period)
        eligible_pension_income= pension_income * head_or_spouse
        total_allowed_pension_exclusion = min_(p.maximum_amount, eligible_pension_income)
        return tax_unit.sum(total_allowed_pension_exclusion)
   
```

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/az/tax/income/net_income/az_pension_exclusion.py]: taxable_

File: policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    href: https://azdor.gov/sites/default/files/2023-03/FORMS_INDIVIDUAL_2022_140BOOKLET.pdf#page=18
+  - title: 2021 Arizona Form 140 Resident Personal Income Tax Booklet
+    href: https://azdor.gov/sites/default/files/2023-03/FORMS_INDIVIDUAL_2021_140BOOKLET.pdf#page=23
+  - title: 43-1022. Subtractions from Arizona gross income
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml:11]: add specific section of the law

File: policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+description: Arizona allows an exclusion of taxable pension income to get its net income that is capped at this amount.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml:1]:
```suggestion
description: Arizona limits the benefits, annuities and pensions to this maximum amount under the pension exclusion.
```

File: policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+metadata:
+  label: Arizona maximum pension exclusion amount
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml:3]:
```suggestion
  label: Arizona pension exclusion max amount
```

File: policyengine_us/variables/gov/states/az/tax/income/net_income/az_pension_exclusion.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+from policyengine_us.model_api import *
+
+
+class az_pension_exclusion(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "Exclusion Arizona State or Local Government Pensions"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/net_income/az_pension_exclusion.py:7]:
```suggestion
    label = "Arizona Pension Exclusion"
```

File: changelog_entry.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+- bump: minor
+  changes:
+    added:
+    - Arizona exclusion for state or local government pensions
```
Comment (PavelMakarchuk) [inline changelog_entry.yaml:4]:
```suggestion
    - Arizona pension exclusion.
```

File: policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+metadata:
+  label: Arizona pension exclusion max amount
+  period: year
+  unit: currency-USD
+  reference:
+  - title: 2022 Arizona Form 140 Resident Personal Income Tax Booklet
+    href: https://azdor.gov/sites/default/files/2023-03/FORMS_INDIVIDUAL_2022_140BOOKLET.pdf#page=18
+  - title: 2021 Arizona Form 140 Resident Personal Income Tax Booklet
+    href: https://azdor.gov/sites/default/files/2023-03/FORMS_INDIVIDUAL_2021_140BOOKLET.pdf#page=23
+  - title: 43-1022. Subtractions from Arizona gross income
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml:11]:
```suggestion
  - title: 43-1022. Subtractions from Arizona gross income (2)
```

File: policyengine_us/variables/gov/states/az/tax/income/net_income/az_pension_exclusion.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+        "https://www.azleg.gov/ars/43/01022.htm"
+    )
+    defined_for = StateCode.AZ
+
+    def formula(tax_unit, period, parameters):
+        p = parameters(period).gov.states.az.tax.income.pension_exclusion
+        person = tax_unit.members
+        is_head = person("is_tax_unit_head", period)
+        is_spouse = person("is_tax_unit_spouse", period)
+        pension_income = person("taxable_public_pension_income", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/net_income/az_pension_exclusion.py:22]: @MaxGhenis would the following qualify as public pension income or are there more variable inputs Yizhou can use?

image

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: make a subtractions.yaml and include this exclusion in the file

File: policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml [hunk 10]
```diff
@@ -0,0 +1,16 @@
+description: Arizona limits the benefits, annuities and pensions to this maximum amount under the pension exclusion.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml:1]:
```suggestion
description: Arizona caps the pension exclusion at this amount.
```

File: policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+metadata:
+  label: Arizona pension exclusion max amount
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml:3]:
```suggestion
  label: Arizona pension exclusion cap
```

File: policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/az/tax/income/pension_exclusion/maximum_amount.yaml:1]: Move to subtractions folder

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
Also please rename the files and variables from `pension` to `public_pension`

**Comment by phywzc** [comment]
Updated the files in response to feedback.

File: policyengine_us/variables/gov/states/az/tax/income/net_income/az_public_pension_exclusion.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/net_income/az_public_pension_exclusion.py:1]: change the `net_income` folder to `subtractions`

File: policyengine_us/tests/policy/baseline/gov/states/az/tax/income/net_income/az_public_pension_exclusion.yaml
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/az/tax/income/net_income/az_public_pension_exclusion.yaml:1]: change the `net_income` folder to `subtractions`

**Review by PavelMakarchuk — APPROVED** [review]

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: git pull upstream master

**Review by MaxGhenis — APPROVED** [review]

**Pull request merged**